tests/integration/helpers.py

```diff
@@ -61,3 +67,127 @@ async def update_juju_secret(
+        # output.stdout.decode(), output.stderr.decode(), output.returncode
+        result = output.stdout.decode()
+        logger.info(f"Command: {command}")
+        logger.info(f"Secrets for namespace: {namespace}")
+        logger.info(f"Request secret: {secret_name}")
+        logger.info(f"results: {str(result)}")
+        secrets = json.loads(result)
+        data = {}
+        for secret in secrets["items"]:
+            name = secret["metadata"]["name"]
+            logger.info(f"\t secretName: {name}")
+            if name == secret_name:
+                data = {}
+                if "data" in secret:
+                    data = secret["data"]
+        return data
+    except subprocess.CalledProcessError as e:
+        return e.stdout.decode(), e.stderr.decode(), e.returncode
+
+
+async def juju_sleep(ops: OpsTest, app_name: str, time: int):
+    await ops.model.wait_for_idle(
+        apps=[
+            app_name,
+        ],
+        idle_period=time,
+        timeout=300,
+    )
+
+
+def setup_s3_bucket_for_sch_server(endpoint_url: str, aws_access_key: str, aws_secret_key: str):
+    config = Config(connect_timeout=60, retries={"max_attempts": 0})
+    session = boto3.session.Session(
+        aws_access_key_id=aws_access_key, aws_secret_access_key=aws_secret_key
+    )
+    s3 = session.client("s3", endpoint_url=endpoint_url, config=config)
+    # delete test bucket and its content if it already exist
+    buckets = s3.list_buckets()
+    for bucket in buckets["Buckets"]:
+        bucket_name = bucket["Name"]
+        if bucket_name == BUCKET_NAME:
+            logger.info(f"Deleting bucket: {bucket_name}")
+            objects = s3.list_objects_v2(Bucket=BUCKET_NAME)["Contents"]
+            objs = [{"Key": x["Key"]} for x in objects]
+            s3.delete_objects(Bucket=BUCKET_NAME, Delete={"Objects": objs})
+            s3.delete_bucket(Bucket=BUCKET_NAME)
+
+    logger.info("create bucket in minio")
+    for i in range(0, 30):
+        try:
+            s3.create_bucket(Bucket=BUCKET_NAME)
+            break
+        except Exception as e:
+            if i >= 30:
+                logger.error(f"create bucket failed....exiting....\n{str(e)}")
+                raise
+            else:
+                logger.warning(f"create bucket failed....retrying in 10 secs.....\n{str(e)}")
+                sleep(10)
+                continue
+
+    s3.put_object(Bucket=BUCKET_NAME, Key=("spark-events/"))
+    logger.debug(s3.list_buckets())
+
+
+async def run_action(
+    ops_test: OpsTest, action_name: str, params: Dict[str, str], app_name: str, num_unit=0
+) -> Any:
+    """Use the charm action to start a password rotation."""
+    action = await ops_test.model.units.get(f"{app_name}/{num_unit}").run_action(
+        action_name, **params
+    )
+    password = await action.wait()
+    return password.results
+
+
+def flatten(map: MutableMapping, parent: str = "", separator: str = ".") -> dict[str, str]:
```

Comment:
Ah, I remember this!

This is so embarassing we have to do this to assert the results of the action -- while the results of the actions were not supposed to be formatted in the nested way in the first place. I think we should also maybe put a ticket in the backlog to fix the output format of the action. (maybe encode the entire stuff as a string or something).

Reply:
I will add ticket